homeassistant/components/modbus/modbus.py

```diff
@@ -187,166 +192,112 @@ def setup(self, hass):
             self._log_error(exception_error, error_state=False)
             return
 
-        # Connect device
-        self.connect()
+        async with self._lock:
+            await self.hass.async_add_executor_job(self._pymodbus_connect)
 
         # Start counting down to allow modbus requests.
         if self._config_delay:
-            self._cancel_listener = call_later(hass, self._config_delay, self.end_delay)
+            self._async_cancel_listener = async_call_later(
+                self.hass, self._config_delay, self.async_end_delay
+            )
 
-    def end_delay(self, args):
+    async def async_end_delay(self, args):
```

Comment:
This can be a callback since we don't await inside.
```suggestion
    @callback
    def async_end_delay(self, args):
```

Reply:
I renamed the function as well to callback_end_delay.
Done